File: src/gates.rs
//! Single-qubit gate matrices and identity operators.
//!
//! Provides the standard Pauli matrices (X, Y, Z) and identity matrices
//! used as building blocks throughout the simulation.

use faer::{Mat, c64};

/// Returns the 2×2 Pauli X matrix.
///
/// ```text
/// X = [[0, 1],
///      [1, 0]]
/// ```
pub fn pauli_x() -> Mat<c64> {
    Mat::from_fn(2, 2, |i, j| match (i, j) {
        (0, 1) | (1, 0) => c64::new(1.0, 0.0),
        _ => c64::new(0.0, 0.0),
    })
}

/// Returns the 2×2 Pauli Y matrix.
///
/// ```text
/// Y = [[ 0, -i],
///      [ i,  0]]
/// ```
pub fn pauli_y() -> Mat<c64> {
    Mat::from_fn(2, 2, |i, j| match (i, j) {
        (0, 1) => c64::new(0.0, -1.0),
        (1, 0) => c64::new(0.0, 1.0),
        _ => c64::new(0.0, 0.0),
    })
}

/// Returns the 2×2 Pauli Z matrix.
///
/// ```text
/// Z = [[1,  0],
///      [0, -1]]
/// ```
pub fn pauli_z() -> Mat<c64> {
    Mat::from_fn(2, 2, |i, j| match (i, j) {
        (0, 0) => c64::new(1.0, 0.0),
        (1, 1) => c64::new(-1.0, 0.0),
        _ => c64::new(0.0, 0.0),
    })
}

/// Returns the 2×2 identity matrix.
pub fn identity_2() -> Mat<c64> {
    Mat::from_fn(2, 2, |i, j| {
        if i == j { c64::new(1.0, 0.0) } else { c64::new(0.0, 0.0) }
    })
}

/// Returns the 2ⁿ × 2ⁿ identity matrix.
pub fn identity_n(n: usize) -> Mat<c64> {
    let dim = 1 << n;
    Mat::from_fn(dim, dim, |i, j| {
        if i == j { c64::new(1.0, 0.0) } else { c64::new(0.0, 0.0) }
    })
}

File: python/gates.py
"""Single-qubit gate matrices and identity operators.

Provides the standard Pauli matrices (X, Y, Z) and identity matrices
used as building blocks throughout the simulation.
"""

import numpy as np


def pauli_x() -> np.ndarray:
    """Returns the 2x2 Pauli X matrix.

    X = [[0, 1],
         [1, 0]]
    """
    return np.array([[0, 1], [1, 0]], dtype=complex)


def pauli_y() -> np.ndarray:
    """Returns the 2x2 Pauli Y matrix.

    Y = [[ 0, -i],
         [ i,  0]]
    """
    return np.array([[0, -1j], [1j, 0]], dtype=complex)


def pauli_z() -> np.ndarray:
    """Returns the 2x2 Pauli Z matrix.

    Z = [[1,  0],
         [0, -1]]
    """
    return np.array([[1, 0], [0, -1]], dtype=complex)


def identity_2() -> np.ndarray:
    """Returns the 2x2 identity matrix."""
    return np.eye(2, dtype=complex)


def identity_n(n: int) -> np.ndarray:
    """Returns the 2^n x 2^n identity matrix."""
    return np.eye(1 << n, dtype=complex)
